turbopack/crates/turbo-tasks-fs/src/watcher.rs

```diff
@@ -250,11 +285,40 @@ impl DiskWatcher {
         let mut batched_new_paths = FxHashSet::default();
 
         'outer: loop {
-            let mut event = rx.recv().or(Err(TryRecvError::Disconnected));
+            let mut event_result = rx.recv().or(Err(TryRecvError::Disconnected));
+            // this inner loop batches events using `try_recv`
             loop {
-                match event {
-                    Ok(Ok(notify::Event { kind, paths, .. })) => {
-                        let paths: Vec<PathBuf> = paths
+                match event_result {
+                    Ok(Ok(event)) => {
+                        // TODO: We might benefit from some user-facing diagnostics if it rescans
+                        // occur frequently (i.e. more than X times in Y minutes)
+                        if event.need_rescan() {
+                            let _lock = inner.invalidation_lock.blocking_write();
+
+                            #[cfg(not(any(target_os = "macos", target_os = "windows")))]
```

Comment:
When not in recursive watching mode, the rescan event would only apply to a single directory and we would have to invalidate everything...

Reply:
At least on Linux, rescan events come with an empty paths list. I'm adding a comment to explain that.

```
Event { kind: Other, paths: [], attr:tracker: None, attr:flag: Some(Rescan), attr:info: None, attr:source: None }
```

Reply:
Yes, but that event happens on a specific watcher instance, don't it?

Reply:
We have a single `notify::Watcher` instance for the entire process that's configured to watch multiple directories.

Yes, with some changes to `notify-rs` or if we created one `notify::Watcher` per watched item (maybe expensive on some platforms?), we might be able to solve this. I don't think that's worth it if we assume that rescans are rare.